step-ca/templates.mdx

```diff
@@ -150,6 +152,95 @@ You can also import values
 from the `"templateData"` object
 in your provisioner's [configuration block](#configuring-the-ca-to-use-templates).
 
+### ASN.1 Values
+
+The X.509 templating system includes functions for ASN.1 encoding in a template.
+This can be useful for populating custom OID `extensions`:
+
+```
+"extensions": [
+  {"id": "1.2.3.4", "critical": true, "value": {{ asn1Enc "int:3848281" | toJson }}}
+]
+```
+
+When applied to template variables, these functions enable dynamic OID values in your certificate:
+
+```
+{
+	"subject": {{ toJson .Subject }},
+	"sans": {{ toJson .SANs }},
+{{- if typeIs "*rsa.PublicKey" .Insecure.CR.PublicKey }}
+	"keyUsage": ["keyEncipherment", "digitalSignature"],
+{{- else }}
+	"keyUsage": ["digitalSignature"],
+{{- end }}
+	"extKeyUsage": ["serverAuth", "clientAuth"],
+	"extensions": [
+		{"id": "1.2.3.5", "value": {{ asn1Marshal .AuthorizationCrt.NotAfter | toJson }}},
+		{"id": "1.2.3.6", "value": {{ asn1Seq (asn1Enc (first .Insecure.CR.DNSNames)) (asn1Enc "int:123456") | toJson }}},
+		{"id": "1.2.3.7", "value": {{ asn1Set (asn1Marshal (first .Insecure.CR.DNSNames) "utf8") (asn1Enc "int:123456") | toJson }}}
+	]
+}
+```
+
+All of the functions here return base64.
+
+To make that possible, `asn1Seq` and `asn1Set` will decode a list of base64 values before putting them into an ASN.1 set or sequence, which is then re-encoded to base64.
+
+For example:
+
+```
+{"id": "1.2.3.4", "value": {{ asn1Seq (asn1Enc "YubiKey") (asn1Enc "int:123456") | toJson }}}
+```
+
+1. `asn1Enc "YubiKey"` will encode "YubiKey" as an ASN.1 Printable string, and return the base64
+2. `asn1Enc "int:123456"` will encode 123 as an ASN.1 Integer, and return the base64
+3. `ans1Seq` wil get all the base64 strings, decode them, create an ASN.1 SEQENCE and return the base64 of the SEQUENCE.
+4. Finally, `toJson` returns a JSON string representation of the base64
+
+#### asn1Enc
+
+Encodes a string into an ASN.1 value.
+A data type can be supplied inline.
+
+```
+asn1Enc "int:123"
+asn1Enc "foo"
+asn1Enc .Subject.CommonName
+```
+
+#### asn1Seq
+
+Creates an ASN.1 `SEQUENCE` of values.
+
+```
+asn1Seq (asn1Enc "foo") (asn1Enc "int:123") ...
+```
+
+#### asn1Set
+
+Creates an ASN.1 `SET` of values.
+
+```
+asn1Set (asn1Enc "foo") (asn1Enc "int:123") ...
+```
+
+#### asn1Marshal
+
+The intention of `asn1Marshal` is to be able to marshal variable data.
+For example, if you want to encode the Not After value in an X5C authorization certificate,
+without converting it to a string first, `asn1Marshal .AuthorizationCrt.NotAfter` will do the trick.
+
+`asn1Marshal` imitates Go's [`asn1.MarshalWithParams`](https://pkg.go.dev/encoding/asn1#MarshalWithParams).
+
+```
+asn1Marshal .Token.iss "printable"
+asn1Marshal .AuthorizationCrt.NotAfter "utc"
+```
+
+The second parameter ("printable", "utc") is the ASN.1 data type.
+The data types are listed in the [`encoding/asn1` package documentation](https://pkg.go.dev/encoding/asn1)
```

Comment:
The second parameter is not required and it will use Go's default transformation.